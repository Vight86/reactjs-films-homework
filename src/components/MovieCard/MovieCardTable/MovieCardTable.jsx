import React from 'react';
import propTypes from 'prop-types';
import Button from '../../Button/index';
import Title from '../../Title/index';
import GenreList from '../../GenreList/index';
import Rating from '../../Rating/index';
import Cover from '../../Cover/index';
import MovieSynopsis from '../../MovieSynopsis/index';
import Modal from '../../Modal/index';
import styles from './MovieCardTable.scss';

const MovieCardTable = ({ data }) => {
  const {
    movie,
    genres,
    movieTrailerKey,
    isInfoShown,
    handleShowInfo,
    isModalOpened,
    handleModalToggle,
  } = data;

  const {
    id, title, voteAverage, posterPath, backdropPath, genreIds, overview,
  } = movie;

  const genreList = genreIds.map((genreId) => genres.find((item) => item.id === genreId).name);

  return (
    <article className={styles.movieCard}>
      <section className={styles.poster}>
        <Cover className="inCard">
          {posterPath && <img src={`https://image.tmdb.org/t/p/w780${posterPath}`} alt={title} />}
        </Cover>
        <div className={styles.viewInfo}>
          <Button className="playButton" onClick={() => handleModalToggle(id, isModalOpened)} />
          <span className={styles.watchNowText}>Watch Now</span>
          <Button className="primary" onClick={handleShowInfo}>View Info</Button>
        </div>
      </section>
      <section className={styles.generalInfo}>
        <Title className="secondary_table">{title}</Title>
        <Rating className="secondary_table">{voteAverage}</Rating>
        <GenreList className="secondary">{genreList}</GenreList>
        <MovieSynopsis className="tertiary">{overview}</MovieSynopsis>
      </section>
      <section
        className={`${styles.infoBlock} ${isInfoShown === true ? styles.infoBlock_active : ''}`.trim()}
      >
        <Cover className="backdropCover">
          {backdropPath && <img src={`https://image.tmdb.org/t/p/w780${backdropPath}`} alt={title} />}
        </Cover>
        <Button className="closeButton_inMovieCard" onClick={handleShowInfo}>&times;</Button>
        <section className={styles.generalInfo}>
          <Title className="secondary_table-light">{title}</Title>
          <Rating className="secondary_table">{voteAverage}</Rating>
          <GenreList className="secondary">{genreList}</GenreList>
        </section>
        <MovieSynopsis className="secondary_table">{overview}</MovieSynopsis>
        <Button className="secondary" onClick={() => handleModalToggle(id, isModalOpened)}>Watch Now</Button>
      </section>
      <Modal
        isModalOpened={isModalOpened}
        movieTrailerKey={movieTrailerKey}
        handleModalToggle={handleModalToggle}
      />
    </article>
  );
};

MovieCardTable.propTypes = {
  data: propTypes.shape({
    movie: propTypes.shape({
      adult: propTypes.bool,
      backdropPath: propTypes.string,
      genreIds: propTypes.arrayOf(propTypes.number),
      id: propTypes.number,
      originalLanguage: propTypes.string,
      originalTitle: propTypes.string,
      overview: propTypes.string,
      popularity: propTypes.number,
      posterPath: propTypes.string,
      releaseDate: propTypes.string,
      title: propTypes.string,
      video: propTypes.bool,
      voteAverage: propTypes.number,
      voteCount: propTypes.number,
    }),
    genres: propTypes.arrayOf(propTypes.object),
    movieTrailerKey: propTypes.string,
    isInfoShown: propTypes.bool,
    handleShowInfo: propTypes.func,
    isModalOpened: propTypes.bool,
    handleModalToggle: propTypes.func,
  }).isRequired,
};

export default MovieCardTable;
