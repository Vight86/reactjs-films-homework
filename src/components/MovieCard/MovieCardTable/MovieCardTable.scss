@import '../../../utils/variables';

.viewInfo {
  position: absolute;
  top: 0%;
  left: 0;
  width: 100%;
  height: 100%;
  padding-top: 21%;
  background-color: transparentize($color: $color-dark, $amount: 0.16);
  opacity: 0;
  text-align: center;
  transform: scale(1.25);
  transition: all 0.3s;
}

.movieCard {
  position: relative;
  display: flex;
  overflow: hidden;
  margin-bottom: 4em;
  background-color: $color-light;
  border-radius: 0.6em;
  box-shadow: 0 10px 40px 0 rgba(0, 0, 0, 0.04);
  font-size: 1rem;

  &:hover {
    .viewInfo {
      opacity: 1;
      transform: scale(1);
    }
  }

  @media screen and (max-width: $bp-xxsmall) {
    flex-direction: column;
  }
}

.poster {
  position: relative;
  overflow: hidden;
  width: 30%;
  flex-shrink: 0;

  @media screen and (max-width: $bp-xxsmall) {
    width: 100%;
  }
}

.watchNowText {
  display: block;
  margin-bottom: 11%;
  color: $bg-color;
  font-size: 2.8em;
  letter-spacing: 0.04em;
}

.generalInfo {
  position: relative;
  display: block;
  width: 70%;
  box-sizing: border-box;
  padding: 5% 8% 5% 8%;
  font-size: 1.4rem;

  @media screen and (max-width: $bp-xxsmall) {
    width: 100%;
  }
}

.infoBlock {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: $color-tertiary;
  background-position: center;
  background-repeat: no-repeat;
  background-size: auto, cover;
  opacity: 0;
  transition: all 0.5s;
  visibility: hidden;

  & .generalInfo {
    z-index: 10;
    width: 100%;
    padding: 85px 8% 0 8%;
  }

  &::after {
    position: absolute;
    z-index: 7;
    top: 0;
    left: 0;
    display: block;
    width: 100%;
    height: 100%;
    background-color: transparentize($color: $color-dark, $amount: 0.36);
    content: '';
  }

  &_active {
    opacity: 1;
    visibility: visible;
  }
}
