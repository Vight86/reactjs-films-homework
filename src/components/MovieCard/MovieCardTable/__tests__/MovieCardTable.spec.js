import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import MovieCardTable from '../index';

const mockHandleModalToggle = jest.fn(() => 'tested');

const mockData = {
  movie: {
    adult: false,
    backdropPath: '/test.jpg',
    genreIds: [1, 2],
    id: 1,
    originalLanguage: 'en',
    originalTitle: 'test',
    overview: 'test',
    popularity: 1,
    posterPath: '/test.jpg',
    releaseDate: '1',
    title: 'test',
    video: false,
    voteAverage: 1,
    voteCount: 1,
  },
  genres: [{
    id: 1,
    name: 'Action',
  },
  {
    id: 2,
    name: 'SciFi',
  }],
  movieTrailerKey: 'test',
  isInfoShown: false,
  handleShowInfo() { },
  isModalOpened: false,
  handleModalToggle() {
    mockHandleModalToggle();
  },
};

let renderer;
let result;

beforeEach(() => {
  renderer = new ShallowRenderer();
  renderer.render(<MovieCardTable data={mockData} />);
  result = renderer.getRenderOutput();
});

afterEach(() => {
  renderer = null;
  result = null;
});

describe('render MovieCardTable component', () => {
  it('render component with props', () => {
    expect(result).toMatchSnapshot();
  });

  it('render playButton click', () => {
    const playButton = result.props.children[0].props.children[1].props.children[0];
    playButton.props.onClick();
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('render watchNow button click', () => {
    const watchNowButton = result.props.children[2].props.children[4];
    watchNowButton.props.onClick();
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('render viewInfoButton button click', () => {
    let infoBlock = result.props.children[2].props.className;
    expect(infoBlock).toMatch(/infoBlock/);
    mockData.isInfoShown = true;
    renderer.render(<MovieCardTable data={mockData} />);
    result = renderer.getRenderOutput();
    infoBlock = result.props.children[2].props.className;
    expect(infoBlock).toMatch(/infoBlock_active/);
  });
});
