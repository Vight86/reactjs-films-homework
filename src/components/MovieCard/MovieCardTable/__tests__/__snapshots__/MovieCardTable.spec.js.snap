// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render MovieCardTable component render component with props 1`] = `
<article
  className="movieCard"
>
  <section
    className="poster"
  >
    <Cover
      className="inCard"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <div
      className="viewInfo"
    >
      <Button
        className="playButton"
        onClick={[Function]}
      >
        
      </Button>
      <span
        className="watchNowText"
      >
        Watch Now
      </span>
      <Button
        className="primary"
        onClick={[Function]}
      >
        View Info
      </Button>
    </div>
  </section>
  <section
    className="generalInfo"
  >
    <Title
      className="secondary_table"
    >
      test
    </Title>
    <Rating
      className="secondary_table"
      withStars={false}
    >
      1
    </Rating>
    <GenreList
      className="secondary"
      order={null}
    >
      Action
      SciFi
    </GenreList>
    <MovieSynopsis
      className="tertiary"
      isShown={false}
    >
      test
    </MovieSynopsis>
  </section>
  <section
    className="infoBlock"
  >
    <Cover
      className="backdropCover"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <Button
      className="closeButton_inMovieCard"
      onClick={[Function]}
    >
      ×
    </Button>
    <section
      className="generalInfo"
    >
      <Title
        className="secondary_table-light"
      >
        test
      </Title>
      <Rating
        className="secondary_table"
        withStars={false}
      >
        1
      </Rating>
      <GenreList
        className="secondary"
        order={null}
      >
        Action
        SciFi
      </GenreList>
    </section>
    <MovieSynopsis
      className="secondary_table"
      isShown={false}
    >
      test
    </MovieSynopsis>
    <Button
      className="secondary"
      onClick={[Function]}
    >
      Watch Now
    </Button>
  </section>
  <Modal
    handleModalToggle={[Function]}
    isModalOpened={false}
    movieTrailerKey="test"
  />
</article>
`;

exports[`render MovieCardTable component render playButton click 1`] = `
<article
  className="movieCard"
>
  <section
    className="poster"
  >
    <Cover
      className="inCard"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <div
      className="viewInfo"
    >
      <Button
        className="playButton"
        onClick={[Function]}
      >
        
      </Button>
      <span
        className="watchNowText"
      >
        Watch Now
      </span>
      <Button
        className="primary"
        onClick={[Function]}
      >
        View Info
      </Button>
    </div>
  </section>
  <section
    className="generalInfo"
  >
    <Title
      className="secondary_table"
    >
      test
    </Title>
    <Rating
      className="secondary_table"
      withStars={false}
    >
      1
    </Rating>
    <GenreList
      className="secondary"
      order={null}
    >
      Action
      SciFi
    </GenreList>
    <MovieSynopsis
      className="tertiary"
      isShown={false}
    >
      test
    </MovieSynopsis>
  </section>
  <section
    className="infoBlock"
  >
    <Cover
      className="backdropCover"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <Button
      className="closeButton_inMovieCard"
      onClick={[Function]}
    >
      ×
    </Button>
    <section
      className="generalInfo"
    >
      <Title
        className="secondary_table-light"
      >
        test
      </Title>
      <Rating
        className="secondary_table"
        withStars={false}
      >
        1
      </Rating>
      <GenreList
        className="secondary"
        order={null}
      >
        Action
        SciFi
      </GenreList>
    </section>
    <MovieSynopsis
      className="secondary_table"
      isShown={false}
    >
      test
    </MovieSynopsis>
    <Button
      className="secondary"
      onClick={[Function]}
    >
      Watch Now
    </Button>
  </section>
  <Modal
    handleModalToggle={[Function]}
    isModalOpened={false}
    movieTrailerKey="test"
  />
</article>
`;

exports[`render MovieCardTable component render watchNow button click 1`] = `
<article
  className="movieCard"
>
  <section
    className="poster"
  >
    <Cover
      className="inCard"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <div
      className="viewInfo"
    >
      <Button
        className="playButton"
        onClick={[Function]}
      >
        
      </Button>
      <span
        className="watchNowText"
      >
        Watch Now
      </span>
      <Button
        className="primary"
        onClick={[Function]}
      >
        View Info
      </Button>
    </div>
  </section>
  <section
    className="generalInfo"
  >
    <Title
      className="secondary_table"
    >
      test
    </Title>
    <Rating
      className="secondary_table"
      withStars={false}
    >
      1
    </Rating>
    <GenreList
      className="secondary"
      order={null}
    >
      Action
      SciFi
    </GenreList>
    <MovieSynopsis
      className="tertiary"
      isShown={false}
    >
      test
    </MovieSynopsis>
  </section>
  <section
    className="infoBlock"
  >
    <Cover
      className="backdropCover"
    >
      <img
        alt="test"
        src="https://image.tmdb.org/t/p/w780/test.jpg"
      />
    </Cover>
    <Button
      className="closeButton_inMovieCard"
      onClick={[Function]}
    >
      ×
    </Button>
    <section
      className="generalInfo"
    >
      <Title
        className="secondary_table-light"
      >
        test
      </Title>
      <Rating
        className="secondary_table"
        withStars={false}
      >
        1
      </Rating>
      <GenreList
        className="secondary"
        order={null}
      >
        Action
        SciFi
      </GenreList>
    </section>
    <MovieSynopsis
      className="secondary_table"
      isShown={false}
    >
      test
    </MovieSynopsis>
    <Button
      className="secondary"
      onClick={[Function]}
    >
      Watch Now
    </Button>
  </section>
  <Modal
    handleModalToggle={[Function]}
    isModalOpened={false}
    movieTrailerKey="test"
  />
</article>
`;
